tempesta_fw/sock_frontend.c

```diff
@@ -33,132 +34,152 @@
 #include <linux/net.h>
 #include <net/inet_sock.h>
 
-#include "tempesta.h"
+#include "cfg.h"
 #include "connection.h"
 #include "filter.h"
 #include "http.h"
 #include "log.h"
-
 #include "sync_socket.h"
+#include "tempesta.h"
 
+#define LISTEN_SOCK_BACKLOG_LEN 1024
+#define LISTEN_SOCKS_MAX 8
+
+static struct socket *listen_socks[LISTEN_SOCKS_MAX];
 static unsigned int listen_socks_n = 0;
-static SsProto *protos;
+
+static SsProto protos[LISTEN_SOCKS_MAX];
+
+#define FOR_EACH_SOCK(sock, i) \
+	for (i = 0;  (sock = listen_socks[i], i < listen_socks_n);  ++i)
 
 /**
- * Create a listening front-end socket.
+ * Parse IP address, create a socket and bind it with the address,
+ * but not yet start listening.
  */
 static int
-__open_listen_socket(SsProto *proto, void *addr)
+add_listen_sock(TfwCfgSpec *cs, TfwCfgEntry *ce)
 {
+	TfwAddr addr;
+	const char *addr_str;
 	struct socket *s;
-	unsigned short family = *(unsigned short *)addr;
-	unsigned short sza = family == AF_INET
-			     ? sizeof(struct sockaddr_in)
-			     : sizeof(struct sockaddr_in6);
 	int r;
 
-	r = sock_create_kern(family, SOCK_STREAM, IPPROTO_TCP, &s);
+	if (listen_socks_n == ARRAY_SIZE(listen_socks)) {
+		TFW_ERR("maximum number of listen sockets (%d) is reached",
+			listen_socks_n);
+		return -ENOBUFS;
+	}
+
+	r = tfw_cfg_check_single_val(ce);
 	if (r) {
-		TFW_ERR("Can't create front-end listening socket (%d)\n", r);
+		TFW_ERR("invalid syntax\n");
 		return r;
 	}
 
-	inet_sk(s->sk)->freebind = 1;
-	s->sk->sk_reuse = 1;
-	r = s->ops->bind(s, (struct sockaddr *)addr, sza);
+	addr_str = ce->vals[0];
+	r = tfw_inet_pton(addr_str, &addr);
 	if (r) {
-		TFW_ERR("Can't bind front-end listening socket (%d)\n", r);
-		goto err;
+		TFW_ERR("can't parse IP address: '%s'\n", ce->vals[0]);
+		return r;
 	}
 
-	ss_tcp_set_listen(s, proto);
-	TFW_DBG("Created listening socket %p\n", s->sk);
-
-	/* TODO adjust /proc/sys/net/core/somaxconn */
-	r = s->ops->listen(s, 1024);
+	r = sock_create_kern(addr.sa.sa_family, SOCK_STREAM, IPPROTO_TCP, &s);
 	if (r) {
-		TFW_ERR("Can't listen on front-end socket (%d)\n", r);
-		goto err;
+		TFW_ERR("Can't create socket (err: %d)\n", r);
+		return r;
 	}
 
-	return r;
-err:
-	sock_release(s);
-	return r;
-}
-
-void
-tfw_close_listen_sockets(void)
-{
-	down_read(&tfw_cfg.mtx);
```

Comment:
It seems there is not synchronization with the new configuration subsystem. How two concurrent configuration updates are synchronized?